File: python/app/processing/metrics.py
"""Basic waveform / RH metrics for single-shot and batch computation."""

from __future__ import annotations

from typing import Dict, List, Optional

import numpy as np
from scipy import stats
from scipy.signal import find_peaks

from app.data.gedi_loader import WaveformResult
from app.processing.waveform_ops import ProcessParams, process_waveform


def _finite(y: np.ndarray) -> np.ndarray:
    y = np.asarray(y, dtype=np.float64)
    return y[np.isfinite(y)]


def waveform_energy_centroid(y: np.ndarray) -> float:
    y = np.asarray(y, dtype=np.float64)
    w = np.clip(y - np.nanmin(y), 0, None)
    s = np.nansum(w)
    if s <= 0:
        return float("nan")
    idx = np.arange(y.size, dtype=np.float64)
    return float(np.nansum(idx * w) / s)


def waveform_fwhm(y: np.ndarray) -> float:
    y = np.asarray(y, dtype=np.float64)
    if y.size < 3:
        return float("nan")
    peak = np.nanmax(y)
    base = np.nanmin(y)
    half = base + 0.5 * (peak - base)
    above = np.where(y >= half)[0]
    if above.size == 0:
        return float("nan")
    return float(above[-1] - above[0] + 1)


def count_peaks(y: np.ndarray, prominence_ratio: float = 0.15) -> int:
    y = np.asarray(y, dtype=np.float64)
    if y.size < 5:
        return 0
    span = np.nanmax(y) - np.nanmin(y)
    if span <= 0:
        return 0
    peaks, _ = find_peaks(y, prominence=prominence_ratio * span)
    return int(peaks.size)


def ground_canopy_ratio(y: np.ndarray, ground_frac: float = 0.25) -> Dict[str, float]:
    """Split waveform into upper canopy vs lower (ground-like) energy.

    GEDI waveforms typically record from canopy top downward; last samples
    often contain ground return.
    """
    y = np.asarray(y, dtype=np.float64)
    if y.size < 4:
        return {"canopy_energy": float("nan"), "ground_energy": float("nan"), "gcr": float("nan")}
    y_pos = np.clip(y - np.nanmin(y), 0, None)
    n_g = max(1, int(round(y.size * ground_frac)))
    ground = float(np.nansum(y_pos[-n_g:]))
    canopy = float(np.nansum(y_pos[:-n_g]))
    gcr = ground / canopy if canopy > 0 else float("nan")
    return {"canopy_energy": canopy, "ground_energy": ground, "gcr": gcr}


def compute_waveform_metrics(
    result: WaveformResult,
    params: Optional[ProcessParams] = None,
) -> Dict[str, Optional[float]]:
    """Compute a dictionary of metrics for one shot."""
    out: Dict[str, Optional[float]] = {
        "shot_number": float(result.shot_number),
        "lon": float(result.lon),
        "lat": float(result.lat),
        "elev_top": result.elev_top,
        "elev_bot": result.elev_bot,
        "sensitivity": result.sensitivity,
        "quality": result.quality,
        "degrade": result.degrade,
        "delta_time": result.delta_time,
        "product": result.product,
        "beam": result.beam,
        "energy_total": result.energy_total,
        "selected_algorithm": result.selected_algorithm,
        "n_samples": None,
        "wf_mean": None,
        "wf_std": None,
        "wf_min": None,
        "wf_max": None,
        "wf_energy": None,
        "wf_centroid": None,
        "wf_fwhm": None,
        "wf_skew": None,
        "wf_kurtosis": None,
        "wf_peak_count": None,
        "canopy_energy": None,
        "ground_energy": None,
        "gcr": None,
        "rh25": None,
        "rh50": None,
        "rh75": None,
        "rh98": None,
        "rh100": None,
        "canopy_height": None,
    }

    # Processed waveform preferred for metrics when filters are enabled
    y = None
    if result.has_waveform():
        raw = result.waveform
        if params is not None:
            y = process_waveform(raw, params)["processed"]
        else:
            y = np.asarray(raw, dtype=np.float64)
        yf = _finite(y)
        if yf.size:
            out["n_samples"] = float(y.size)
            out["wf_mean"] = float(np.mean(yf))
            out["wf_std"] = float(np.std(yf))
            out["wf_min"] = float(np.min(yf))
            out["wf_max"] = float(np.max(yf))
            out["wf_energy"] = float(np.nansum(np.clip(y - np.nanmin(y), 0, None)))
            out["wf_centroid"] = waveform_energy_centroid(y)
            out["wf_fwhm"] = waveform_fwhm(y)
            if yf.size >= 3:
                out["wf_skew"] = float(stats.skew(yf))
                out["wf_kurtosis"] = float(stats.kurtosis(yf))
            out["wf_peak_count"] = float(count_peaks(y))
            gcr = ground_canopy_ratio(y)
            out.update(gcr)
        if result.heights is not None and result.elev_top is not None and result.elev_bot is not None:
            # canopy height proxy: highest return - lowest mode elevation
            try:
                out["canopy_height"] = float(result.elev_top - result.elev_bot)
            except Exception:
                out["canopy_height"] = None

    if result.has_rh():
        rh = result.rh
        out["n_samples"] = float(rh.size)
        out["rh25"] = float(np.interp(25, np.arange(rh.size), rh))
        out["rh50"] = float(rh[50]) if rh.size > 50 else float(np.percentile(rh, 50))
        out["rh75"] = float(rh[75]) if rh.size > 75 else float(np.percentile(rh, 75))
        out["rh98"] = float(rh[98]) if rh.size > 98 else float(np.percentile(rh, 98))
        out["rh100"] = float(rh[100]) if rh.size > 100 else float(np.max(rh))
        out["rh_mean"] = float(np.nanmean(rh))
        out["rh_std"] = float(np.nanstd(rh))
        if result.elev_top is not None and result.elev_bot is not None:
            out["canopy_height"] = float(result.elev_top - result.elev_bot)
        # RH profile as pseudo-waveform stats
        out["wf_mean"] = float(np.nanmean(rh))
        out["wf_std"] = float(np.nanstd(rh))
        out["wf_min"] = float(np.nanmin(rh))
        out["wf_max"] = float(np.nanmax(rh))
        out["wf_energy"] = float(np.nansum(np.clip(rh - np.nanmin(rh), 0, None)))
        out["wf_centroid"] = waveform_energy_centroid(rh)
        out["wf_fwhm"] = waveform_fwhm(rh)
        if rh.size >= 3:
            out["wf_skew"] = float(stats.skew(rh))
            out["wf_kurtosis"] = float(stats.kurtosis(rh))
        out["wf_peak_count"] = float(count_peaks(rh))
        gcr = ground_canopy_ratio(rh)
        out.update(gcr)

    return out


METRIC_COLUMNS = [
    "file",
    "product",
    "beam",
    "shot_number",
    "lon",
    "lat",
    "delta_time",
    "sensitivity",
    "quality",
    "degrade",
    "elev_top",
    "elev_bot",
    "canopy_height",
    "n_samples",
    "wf_mean",
    "wf_std",
    "wf_min",
    "wf_max",
    "wf_energy",
    "wf_centroid",
    "wf_fwhm",
    "wf_skew",
    "wf_kurtosis",
    "wf_peak_count",
    "canopy_energy",
    "ground_energy",
    "gcr",
    "rh25",
    "rh50",
    "rh75",
    "rh98",
    "rh100",
    "energy_total",
    "selected_algorithm",
]


def metrics_to_row(file_name: str, metrics: Dict[str, Optional[float]]) -> Dict[str, Optional[float]]:
    row = {c: None for c in METRIC_COLUMNS}
    row["file"] = file_name
    for k, v in metrics.items():
        if k in row:
            row[k] = v
        elif k == "product":
            row["product"] = v
        elif k == "beam":
            row["beam"] = v
    return row

File: cpp/src/processing/Metrics.cpp
#include "processing/Metrics.h"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace proc {

static double nan() { return std::numeric_limits<double>::quiet_NaN(); }

static void seriesStats(const std::vector<double>& y, QMap<QString, double>& m)
{
    std::vector<double> f;
    for (double v : y)
        if (std::isfinite(v))
            f.push_back(v);
    if (f.empty())
        return;
    double mean = std::accumulate(f.begin(), f.end(), 0.0) / f.size();
    double var = 0;
    for (double v : f)
        var += (v - mean) * (v - mean);
    var /= double(f.size());
    auto mm = std::minmax_element(f.begin(), f.end());
    m["wf_mean"] = mean;
    m["wf_std"] = std::sqrt(var);
    m["wf_min"] = *mm.first;
    m["wf_max"] = *mm.second;
    double lo = *mm.first;
    double e = 0;
    for (double v : y)
        e += std::max(0.0, v - lo);
    m["wf_energy"] = e;
    double wsum = 0, csum = 0;
    for (size_t i = 0; i < y.size(); ++i) {
        double w = std::max(0.0, y[i] - lo);
        wsum += w;
        csum += w * double(i);
    }
    m["wf_centroid"] = wsum > 0 ? csum / wsum : nan();
    // FWHM
    double peak = *mm.second;
    double half = lo + 0.5 * (peak - lo);
    int first = -1, last = -1;
    for (size_t i = 0; i < y.size(); ++i) {
        if (y[i] >= half) {
            if (first < 0)
                first = (int)i;
            last = (int)i;
        }
    }
    m["wf_fwhm"] = first >= 0 ? double(last - first + 1) : nan();
    // peaks ~ simple local maxima
    int peaks = 0;
    double span = peak - lo;
    for (size_t i = 1; i + 1 < y.size(); ++i) {
        if (y[i] > y[i - 1] && y[i] >= y[i + 1] && y[i] > lo + 0.15 * span)
            ++peaks;
    }
    m["wf_peak_count"] = peaks;
    // GCR
    size_t n = y.size();
    size_t ng = std::max<size_t>(1, n / 4);
    double ge = 0, ce = 0;
    for (size_t i = 0; i < n; ++i) {
        double w = std::max(0.0, y[i] - lo);
        if (i >= n - ng)
            ge += w;
        else
            ce += w;
    }
    m["ground_energy"] = ge;
    m["canopy_energy"] = ce;
    m["gcr"] = ce > 0 ? ge / ce : nan();
}

QMap<QString, double> computeMetrics(const data::WaveformResult& r)
{
    QMap<QString, double> m;
    m["shot_number"] = double(r.shotNumber);
    m["lon"] = r.lon;
    m["lat"] = r.lat;
    m["elev_top"] = r.elevTop;
    m["elev_bot"] = r.elevBot;
    m["sensitivity"] = r.sensitivity;
    m["quality"] = r.quality;
    m["delta_time"] = r.deltaTime;
    if (std::isfinite(r.elevTop) && std::isfinite(r.elevBot))
        m["canopy_height"] = r.elevTop - r.elevBot;
    if (r.hasWaveform()) {
        m["n_samples"] = double(r.waveform.size());
        seriesStats(r.waveform, m);
    } else if (r.hasRh()) {
        m["n_samples"] = double(r.rh.size());
        seriesStats(r.rh, m);
        auto rhAt = [&](int p) {
            if (r.rh.empty())
                return nan();
            p = std::clamp(p, 0, (int)r.rh.size() - 1);
            return r.rh[p];
        };
        m["rh25"] = rhAt(25);
        m["rh50"] = rhAt(50);
        m["rh75"] = rhAt(75);
        m["rh98"] = rhAt(98);
        m["rh100"] = rhAt(100);
    }
    if (std::isfinite(r.energyTotal))
        m["energy_total"] = r.energyTotal;
    return m;
}

} // namespace proc
